operator/builtin/input/azure/eventhub/event_hub.go

```diff
@@ -0,0 +1,202 @@
+	}
+
+	eventHubInput := &EventHubInput{
+		InputOperator: inputOperator,
+		namespace:     c.Namespace,
+		name:          c.Name,
+		group:         c.Group,
+		connStr:       c.ConnectionString,
+		prefetchCount: c.PrefetchCount,
+		startAtEnd:    startAtEnd,
+		persist: Persister{
+			DB: helper.NewScopedDBPersister(buildContext.Database, c.ID()),
+		},
+	}
+	return []operator.Operator{eventHubInput}, nil
+}
+
+// EventHubInput is an operator that reads input from Azure Event Hub.
+type EventHubInput struct {
+	helper.InputOperator
+	cancel context.CancelFunc
+
+	namespace     string
+	name          string
+	group         string
+	connStr       string
+	prefetchCount uint32
+	startAtEnd    bool
+
+	persist Persister
+	hub     *azhub.Hub
+	wg      sync.WaitGroup
+}
+
+// Start will start generating log entries.
+func (e *EventHubInput) Start() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+
+	if err := e.persist.DB.Load(); err != nil {
+		return err
+	}
+
+	hub, err := azhub.NewHubFromConnectionString(e.connStr, azhub.HubWithOffsetPersistence(&e.persist))
+	if err != nil {
+		return err
+	}
+	e.hub = hub
+
+	runtimeInfo, err := hub.GetRuntimeInformation(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, partitionID := range runtimeInfo.PartitionIDs {
+		go e.poll(ctx, partitionID, hub)
+	}
+
+	return nil
+}
+
+// Stop will stop generating logs.
+func (e *EventHubInput) Stop() error {
+	e.cancel()
+	e.wg.Wait()
+	if err := e.hub.Close(context.Background()); err != nil {
+		return err
+	}
+	e.Infow(fmt.Sprintf("Closed all connections to Azure Event Hub '%s'", e.name))
+	return e.persist.DB.Sync()
+}
+
+// poll starts polling an Azure Event Hub partition id for new events
+func (e *EventHubInput) poll(ctx context.Context, partitionID string, hub *azhub.Hub) error {
+	offsetStr := ""
+	if e.startAtEnd {
+		offset, err := e.persist.Read(e.namespace, e.name, e.group, partitionID)
+		if err != nil {
+			x := fmt.Sprintf("Error while reading offset for partition_id %s, starting at begining", partitionID)
+			e.Errorw(x, zap.Error(err))
+		} else {
+			offsetStr = offset.Offset
+		}
+	}
+
+	var err error
+
+	// start at begining
+	if !e.startAtEnd {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(""),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	// start at end and no offset was found
+	if e.startAtEnd && offsetStr == "" {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithLatestOffset(),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	// start at end and offset exists
+	if e.startAtEnd && offsetStr != "" {
+		_, err = hub.Receive(
+			ctx, partitionID, e.handleEvent, azhub.ReceiveWithStartingOffset(offsetStr),
+			azhub.ReceiveWithPrefetchCount(e.prefetchCount))
+	}
+	if err != nil {
+		return err
+	}
+
+	e.Infow(fmt.Sprintf("Successfully connected to Azure Event Hub '%s' partition_id '%s'", e.name, partitionID))
```

Comment:
This might be a little noisy to print every poll cycle.

Reply:
The method name might be misleading. `poll` runs once per partition. In my environment, this is twice. Poll will call `hub.Receive` which starts a go routine for handling events (this is abstracted in the Azure SDK).

The method name might be better as `startConsumers`. 

```
{"level":"info","timestamp":"2021-04-19T17:10:26.246-0400","message":"Starting stanza agent"}
{"level":"info","timestamp":"2021-04-19T17:10:26.719-0400","message":"Stanza agent started"}
{"level":"info","timestamp":"2021-04-19T17:10:27.238-0400","message":"Successfully connected to Azure Event Hub 'devel' partition_id '0'","operator_id":"$.azure_eventhub.azure_eventhub_input","operator_type":"azure_eventhub_input"}
{"level":"info","timestamp":"2021-04-19T17:10:27.762-0400","message":"Successfully connected to Azure Event Hub 'devel' partition_id '1'","operator_id":"$.azure_eventhub.azure_eventhub_input","operator_type":"azure_eventhub_input"}

```